src/annotator/util/buckets.js

```diff
@@ -0,0 +1,248 @@
+const BUCKET_TOP_THRESHOLD = 115 + BUCKET_NAV_SIZE; // Toolbar
+
+// TODO!! This is an option in the plugin right now
+const BUCKET_GAP_SIZE = 60;
+
+/**
+ * Find the closest valid anchor in `anchors` that is offscreen in the direction
+ * indicated.
+ *
+ * @param {Anchor[]} anchors
+ * @param {'up'|'down'} direction
+ * @return {Anchor|null} - The closest anchor or `null` if no valid anchor found
+ */
+export function findClosestOffscreenAnchor(anchors, direction) {
+  let closestAnchor = null;
+  let closestTop = 0;
+
+  for (let anchor of anchors) {
+    if (!anchor.highlights?.length) {
+      continue;
+    }
+
+    const top = getBoundingClientRect(anchor.highlights).top;
+
+    // Verify that the anchor is offscreen in the direction we're headed
+    if (direction === 'up' && top >= BUCKET_TOP_THRESHOLD) {
+      // We're headed up but the anchor is already below the
+      // visible top of the bucket bar: it's not our guy
+      continue;
+    } else if (
+      direction === 'down' &&
+      top <= window.innerHeight - BUCKET_NAV_SIZE
+    ) {
+      // We're headed down but this anchor is already above
+      // the usable bottom of the screen: it's not our guy
+      continue;
+    }
+
+    if (
+      !closestAnchor ||
+      (direction === 'up' && top > closestTop) ||
+      (direction === 'down' && top < closestTop)
+    ) {
+      // This anchor is either:
+      // - The first anchor we've encountered off-screen in the direction
+      //   we're headed, or
+      // - Closer to the screen than the previous `closestAnchor`
+      closestAnchor = anchor;
+      closestTop = top;
+    }
+  }
+
+  return closestAnchor;
+}
+
+/**
+ * Construct an Array of points representing the positional tops and bottoms
+ * of current anchor highlights. Each anchor whose highlight(s)' bounding
+ * box is onscreen will result in two entries in the `points` Array: one
+ * for the top of the highlight box and one for the bottom
+ *
+ * @param {Anchor[]} anchors
+ * @return {PositionPoints}
+ */
+export function constructPositionPoints(anchors) {
+  const aboveScreenAnchors = new Set();
+  const belowScreenAnchors = new Set();
+  const points = /** @type {PositionPoint[]} */ (new Array());
+
+  for (let anchor of anchors) {
+    if (!anchor.highlights?.length) {
+      continue;
+    }
+
+    const rect = getBoundingClientRect(anchor.highlights);
+
+    if (rect.top < BUCKET_TOP_THRESHOLD) {
+      aboveScreenAnchors.add(anchor);
+    } else if (rect.top > window.innerHeight - BUCKET_NAV_SIZE) {
+      belowScreenAnchors.add(anchor);
+    } else {
+      // Add a point for the top of this anchor's highlight box
+      points.push([rect.top, 1, anchor]);
+      // Add a point for the bottom of this anchor's highlight box
+      points.push([rect.bottom, -1, anchor]);
+    }
+  }
+
+  // Sort onscreen points by pixel position, secondarily by position "type"
+  // (top or bottom of higlight box)
```

Comment:
The `sort` callback doesn't do the "sort by position type" part of this sentence. Are you assuming that the points are already sorted by position before this code is called, and therefore `points` preserves this assuming it uses a stable sort?

Reply:
Stale comment from before my migration changes. Will adjust.

Reply:
Actually, wait, no. How does it not "sort by position type"?  The wording may be confusing here. If the first entry in the Tuple for the compared `Point`s is equal (i.e. they have the same vertical offset position value), this sort function will continue on to the next `Point` value ("type" of point—top or bottom) and sort by _that_. Thus sorting primarily by position, secondarily by type. How would you suggest this comment is worded to be more clear?

Reply:
Oh, right. I forgot that we're dealing with entries that are tuples here. One way to make it clearer would be to make it less generic:

```js
  points.sort(([posA, typeA], [posB, typeB]) => {
    if (posA !== posB) {
      return posA - posB;
    } else {
      return typeA - typeB;
    }
  });
```

This version also has the benefit that it doesn't try to compare `Anchor` objects using the `<` operator if two anchors have the same position and type values. Instead the second comparison will return `0` and so the pre-existing sort order will be preserved.